### PR TITLE
Add Gram baseline

### PR DESCRIPTION
**Pull request by paulnovello**
Implementation of Gram baseline from "Detecting Out-of-Distribution Examples with Gram Matrices" [link](https://proceedings.mlr.press/v119/sastry20a.html)

While implementing this class, we added a new feature to the FeatureExtractor class, which is the ability to preprocess feature maps on the fly, batch-wise, and to directly return processed features as output of `.predict_tensor()`. This is crucial for this method since the computations performed on feature maps are quite intensive and would lead to OOM if not applied batch-wise. 
This feature will be useful for other baselines, e.g. DKNN or mahalanobis (or the planned NMD) to be able to process internal feature maps. In that case, the `postproc_fn` could be some pooling + flatten.

**Important Disclaimer**: Taking the statistics of min/max deviation,
as in the paper raises some problems.

The method often yields a score of zero for some tasks. This is expected since the min/max among the samples of a random variable becomes more and more extreme with the sample size. As a result, computing the min/max over the training set is likely to produce min/max values that are so extreme that none of the in distribution correlations of
the validation set goes beyond these threshold. The worst is that a significant part of ood data does not exceed the thresholds either. This can be alleviated by computing the min/max over a limited number of sample. However, it is counter-intuitive and, in our opinion, not desirable: adding some more information should only improve a method.

Hence, we decided to replace the min/max by the q / 1-q quantile, with q a new parameter of the method. Specifically, instead of the deviation as defined in eq. 3 of the paper, we use the definition

image

With this new deviation, the more points we add, the more accurate the quantile becomes. In addition, the method can be made more or less discriminative by toggling the value of q.

Finally, we found that this approach improved the performance of the baseline in our experiments.

### CODE REVIEW

**Review by y-prudent — COMMENTED** [review]
Really nice work!! A major issue in `_stat` method of `gram.py` need to be fixed.

File: oodeel/extractor/keras_feature_extractor.py [hunk 1]
```diff
@@ -166,6 +172,15 @@ def predict_tensor(self, tensor: TensorType) -> Tuple[List[tf.Tensor], tf.Tensor
...
         logits = features.pop()
+
+        if postproc_fns is not None:
+            if len(postproc_fns) == 1:
+                features = [postproc_fns[0](features)]
+            else:
+                features = [
+                    postproc_fn(feature)
+                    for feature, postproc_fn in zip(features, postproc_fns)
+                ]
```
Comment (y-prudent) [inline oodeel/extractor/keras_feature_extractor.py:183]: This will not work if `len(postproc_fns)` is equal to 1 (so in the case of gram, when there is only one `feature_layer_id`) as features is a list. I suggest to remove the lines from 177 to 179 since `features` and `postproc_fns` are of type List in all case:

```suggestion
        if postproc_fns is not None:
            features = [
                postproc_fn(feature)
                for feature, postproc_fn in zip(features, postproc_fns)
            ]
```

Fixing these few lines is however not enough to make Gram work when there is only 1 feature layer id (check my comment on gram.py that raises an other issue about the type of `tensor_stats`).

File: oodeel/methods/gram.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+            val_split (float): The percentage of fit data to use as validation data for
+                normalization. Default to 0.2.
+        """
+        self.postproc_fns = [
+            self._stat for i in range(len(self.feature_extractor.feature_layers_id))
+        ]
+
+        fit_stats, info = self.feature_extractor.predict(
+            fit_dataset, postproc_fns=self.postproc_fns, return_labels=True
+        )
```
Comment (y-prudent) [inline oodeel/methods/gram.py:112]: If there is only 1 `feature_layer_id`, then `fit_stats` will be a tensor and not a list of tensors at the output of the feature extractor predict function. 

In this case these lines will raise an error:
```python
val_stats = [fit_stat[val_indices] for fit_stat in fit_stats]
fit_stats = [fit_stat[train_indices] for fit_stat in fit_stats]
```

as fit_stat is expected to be a list. The same problem will persist in the `_score_tensor` function where `tensor_stats` is expected to be a list but isn't when there is 1 `feature_layer_id`.

A first solution is to add some conditions on the length of feature_layers_id and to encapsulate the output of the feature extractor when this length is equal to 1 in the code of Gram. An other solution which seems better to me for consistency reasons would be to systematically return a list of tensors in the predict functions of the feature extractors (instead of a tensor when there is 1 output).

File: oodeel/extractor/torch_feature_extractor.py [hunk 3]
```diff
@@ -206,6 +209,16 @@ def predict_tensor(
...
+
+        if postproc_fns is not None:
+            if len(postproc_fns) == 1:
+                features = [postproc_fns[0](features)]
+            else:
+                features = [
+                    postproc_fn(feature)
+                    for feature, postproc_fn in zip(features, postproc_fns)
+                ]
+
```
Comment (y-prudent) [inline oodeel/extractor/torch_feature_extractor.py:221]: Same comment as in KerasFeatureExtractor:

This will not work if `len(postproc_fns)` is equal to 1 (so in the case of gram, when there is only one `feature_layer_id`) as features is a list. I suggest to remove the lines from 214 to 216 since `features` and `postproc_fns` are of type List in all case:

```suggestion
        if postproc_fns is not None:
            features = [
                postproc_fn(feature)
                for feature, postproc_fn in zip(features, postproc_fns)
            ]
```

Fixing these few lines is however not enough to make Gram work when there is only 1 feature layer id (check my comment on gram.py that raises an other issue about the type of `tensor_stats`).

File: tests/tests_tensorflow/methods/test_tf_gram.py
Comment (y-prudent) [inline tests/tests_tensorflow/methods/test_tf_gram.py:1]: I suggest to add a test with `feature_layers_id` of length 1 (that currently would raise an error).

File: tests/tests_torch/methods/test_torch_gram.py
Comment (y-prudent) [inline tests/tests_torch/methods/test_torch_gram.py:1]: I suggest to add a test with `feature_layers_id` of length 1 (that currently would raise an error).

**Review by y-prudent — CHANGES_REQUESTED** [review]
Some remarks concerning the failing case where `feature_layers_id` has a length of 1.

File: oodeel/methods/gram.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+        fm_s = feature_map.shape
+        stat = []
+        for p in self.orders:
+            feature_map_p = feature_map**p
+            # construct the Gram matrix
+            if len(fm_s) == 2:
+                # prevents from computing a scalar per batch for layers of shape (1,)
+                feature_map_p = self.op.einsum("bi,bj->b", feature_map_p, feature_map_p)
+                feature_map_p = self.op.unsqueeze(feature_map_p, -1)
+                feature_map_p = self.op.unsqueeze(feature_map_p, -1)
```
Comment (y-prudent) [inline oodeel/methods/gram.py:250]: Here the statistic for a dense layer behaves as the L2 norm of the activation which provides little information. Instead, I suggest building a gram matrix analogously to the convolutional case, where the feature matrix of a sample $F_l(D)$ has a shape [dim_layer, 1]. In practice, this seems to give better results for OOD detection from an MLP.

 ```suggestion
            if len(fm_s) == 2:
                # build gram matrix for feature map of shape [dim_dense_layer, 1]
                feature_map_p = self.op.einsum("bi,bj->bij", feature_map_p, feature_map_p)
```

**Review by y-prudent — APPROVED** [review]

**Comment by y-prudent** [comment]
LGTM
